api/v1alpha4/cluster_types.go

```diff
@@ -60,6 +60,72 @@ type ClusterSpec struct {
 	// for provisioning infrastructure for a cluster in said provider.
 	// +optional
 	InfrastructureRef *corev1.ObjectReference `json:"infrastructureRef,omitempty"`
+
+	// This encapsulates the topology for the cluster.
+	// NOTE: This feature is alpha; it is required to enable the ClusterTopology
+	// feature gate flag to activate managed topologies support.
+	// +optional
+	Topology *Topology `json:"topology,omitempty"`
```

Comment:
> General question about the PR: does it make master unreleasable?

Master should always be ready for a release.

In order to make it possible while adding ClusterClass and managed topologies on the master branch we are avoiding breaking changes and we are keeping all the new features behind a feature flag, and if the flag is false (which is the default), Cluster API just works as of today with web hooks enforcing you can't set Cluster.Topology or Create/Update ClusterClass types. Also, keeping "strict" isolation between ClusterClass and existing controllers is a driving principle of the next steps of this effort, see e.g. notes on https://github.com/kubernetes-sigs/cluster-api/issues/4933.

Reply:
+1, as long as webhooks are installed (which we always assume) we should be fine. While the fields show up within the CRDs, enabling the feature gate won't do much more than enabling something non functional.

One thing I'd add into the comment for now is `This feature is alpha, highly experimental, and parts of it might still be not implemented`